src/main/scala/com/databricks/sparkdl/ImageUtils.scala

```diff
@@ -35,81 +47,88 @@ private[sparkdl] object ImageUtils {
     val height = ImageSchema.getHeight(rowImage)
     val width = ImageSchema.getWidth(rowImage)
     val channels = ImageSchema.getNChannels(rowImage)
+    val mode = ImageSchema.getMode(rowImage)
     val imageData = ImageSchema.getData(rowImage)
     require(
       imageData.length == height * width * channels,
       s"""| Only one byte per channel is currently supported, got ${imageData.length} bytes for
           | image of size ($height, $width, $channels).
        """.stripMargin
     )
-    val image = new BufferedImage(width, height, BufferedImage.TYPE_3BYTE_BGR)
 
-    var offset, h = 0
-    var r, g, b: Byte = 0
-    while (h < height) {
-      var w = 0
-      while (w < width) {
-        channels match {
-          case 1 =>
-            b = imageData(offset)
-            g = b
-            r = b
-          case 3 =>
-            b = imageData(offset)
-            g = imageData(offset + 1)
-            r = imageData(offset + 2)
-          case _ =>
-            require(false, s"`Channels` must be 1 or 3, got $channels.")
-        }
+    val imageType = openCVModeToImageType.getOrElse(mode,
+      throw new UnsupportedOperationException("Cannot convert row image with  " +
+        s"unsupported OpenCV mode = ${mode} to BufferedImage. Supported OpenCV modes: " +
+        s"${supportedModes.map(ImageSchema.ocvTypes(_)).mkString(", ")}"))
 
-        val color = new Color(r & 0xff, g & 0xff, b & 0xff)
-        image.setRGB(w, h, color.getRGB)
-        offset += channels
-        w += 1
+    val image = new BufferedImage(width, height, imageType)
+    var offset = 0
+    val raster = image.getRaster
+    // NOTE: This code assumes the raw image data in rowImage directly corresponds to the
```

Comment:
Does this change the data compared to the previous way?

Reply:
Good Q, it doesn't change the data layout compared to the previous way. The BufferedImage types supported by our utilities (`TYPE_BYTE_GRAY, TYPE_3_BYTE_BGR, TYPE_4_BYTE_AGBR`) have a fixed BGR(A) channel ordering so this approach should also be consistent across Java versions etc (see [BufferedImage docs](https://docs.oracle.com/javase/7/docs/api/java/awt/image/BufferedImage.html)).

Reply:
Ah I was wondering more about the content, whether what we're doing manually produced the exact same result as `Color` & `Color.getRGB`. The tests look like they maintain the same resizing results as before the code change here but wanted to confirm, since doing anything different to the image can affect the results of applying models to it. (Part of it is, I'm not sure how reasonable this assumption stated here is.)

Reply:
Ahh I see. Actually the resulting BufferedImage may change when feeding grayscale images to DeepImageFeaturizer, but not for three-channel images (our DIF tests only run against three-channel images so they don't catch the change).

The difference for grayscale images is that running `val b = rowImageData(offset); val color = new Color(b, b, b); image.setRGB(x, y, color.getRGB)` as the original code did doesn't necessarily set the grayscale pixel at `(x, y)` to `b`. This happens because Java takes a weighted combination of the (r, g, b) channels when setting the grayscale pixel value. However with the raster approach the pixel at `(x, y)` is actually directly set to `b`. Given that `b` was likely obtained by calling `raster.getSample` (how `ImageSchema.readImages` reads the data of a grayscale image), this seems ok to me.

That being said I could see changes in model behavior being sufficiently undesirable that we wouldn't want to make this change.

Another thing: just realized we no longer validate that the input images to DIF are either 1 or 3 channels, I'm assuming we should add this validation back? The practical implication being that as it stands, this PR allows users to feed 4-channel images to DIF without hitting an error.

Reply:
@tomasatdatabricks mainly I just didn't want to inadvertently release support for four-channel inputs in DIF without adding unit tests :P. Although I suppose we already supported one-channel inputs without specifically testing DIF on one-channel images.

Reply:
ok the current logic looks okay to me. the most important thing is not to change the logic if possible so there is no breaking change for users. i doubt this is a problem at this point. 

for the one- and four- channel input stuff -- if we want any testing, it seems enough to have unit tests to make sure this resizing function returns the correct number of channels given `targetChannels`. Just checking the number of channels seems good enough - we could also check the content of the image if we want to catch any breaking changes in the future.

Reply:
SGTM. Should we add the unit tests in this PR or in a follow-up?